### PR TITLE
build(cmake): skip download if target file exists and hash matched

### PR DESCRIPTION
**Pull request by comicfans**
During build and test neovim, I've found cmake build script ignore manually downloaded deps file even hash matches, this created pain in some restricted environment (proxy not easily used from command line, file can be only downloaded from browser), this patch improve it.

### CODE REVIEW

File: third-party/cmake/DownloadAndExtractFile.cmake [hunk 1]
```diff
@@ -59,42 +59,47 @@ string(REPLACE ";" "-" fname "${fname}")
...
-    list(GET status 1 status_string)
-  endif()
-  if(NOT status_code EQUAL 0)
-    message(FATAL_ERROR "error: downloading '${URL}' failed
-  status_code: ${status_code}
-  status_string: ${status_string}
-  log: ${log}
-")
-  endif()
+if(EXISTS ${file})
```
Comment (seandewar) [inline third-party/cmake/DownloadAndExtractFile.cmake:62]: This will cause the download to never happen if the file wasn't downloaded yet?

Comment (comicfans) [reply third-party/cmake/DownloadAndExtractFile.cmake:63]: silly me, I've fixed it

File: third-party/cmake/DownloadAndExtractFile.cmake [hunk 2]
```diff
@@ -59,44 +59,55 @@ string(REPLACE ";" "-" fname "${fname}")
...
-    execute_process(COMMAND ${CMAKE_COMMAND} -E sleep 10)
-    file(DOWNLOAD ${URL} ${file}
-      ${timeout_args}
-      ${hash_args}
-      STATUS status
-      LOG log)
-    list(GET status 0 status_code)
-    list(GET status 1 status_string)
-  endif()
+set(existing_match FALSE)
```
Comment (justinmk) [inline third-party/cmake/DownloadAndExtractFile.cmake:62]: 1. don't need this extra variable, just set `EXISTING_SHA256` to empty string here . then the `if(NOT existing_match)` condition can be changed to `if(NOT EXISTING_SHA256 STREQUAL "${EXPECTED_SHA256}")`
2. use 2-space indent

Comment (comicfans) [reply third-party/cmake/DownloadAndExtractFile.cmake]: updated with suggestion

File: third-party/cmake/DownloadAndExtractFile.cmake [hunk 3]
```diff
@@ -59,41 +59,48 @@ string(REPLACE ";" "-" fname "${fname}")
...
-      LOG log)
-    list(GET status 0 status_code)
-    list(GET status 1 status_string)
-  endif()
+set(EXISTING_SHA256 "")
+if(EXISTS ${file})
+  file(SHA256 ${file} EXISTING_SHA256)
+endif()
+
+if(NOT EXISTING_SHA256 STREQUAL ${EXPECTED_SHA256})
```
Comment (justinmk) [inline third-party/cmake/DownloadAndExtractFile.cmake:67]: sorry i forgot to mention, is it possible that EXPECTED_SHA256 is empty string? then we should check something like:
```suggestion
if(${EXPECTED_SHA256} STREQUAL "" OR NOT EXISTING_SHA256 STREQUAL ${EXPECTED_SHA256})
```

Comment (justinmk) [reply third-party/cmake/DownloadAndExtractFile.cmake:67]: > is it possible that EXPECTED_SHA256 is empty string?

Looks like no: https://github.com/neovim/neovim/blob/490615612ed5ec587c8023de28db495b3181de30/third-party/cmake/DownloadAndExtractFile.cmake#L13

Comment (comicfans) [reply third-party/cmake/DownloadAndExtractFile.cmake:67]: seems EXPECTED_SHA256 can be set to 'SKIP' to drive some other logic, that's season I don't use it to compare with existing SHA256 in my first PR(although sha256 can't be 'skip')

**Pull request merged**